### PR TITLE
ratelimiter: Don't keep counting after user ran into rate limit

### PR DESCRIPTION
**Pull request by eseliger**
Closes https://github.com/sourcegraph/sourcegraph/issues/51455

This fixes an issue where users would not get additional quota after they got throttled when they tried to hit the API continuously, because we would count the rejected requests, too. 

Example:

Quota: 10, makes 10 requests, gets throttled. Makes another 10 requests, no luck. The internal counter is now at 20. Admin raises quota to 20, user gets 0 additional requests granted as the counter is already at 20.

In a separate PR, I will backport this fix to the LLM proxy limiter as well.

## Test plan

Verified locally that limits are still enforced correctly and that the scenario described above allows 10 more requests instead of throttling again right away.

### CODE REVIEW

**Review by bobheadxi — APPROVED** [review]
Fix LGTM, and okay if it closes out https://github.com/sourcegraph/sourcegraph/issues/51455 - I'm not sure that the behaviour outlined in that issue (reset entirely) is desirable, but it looks like this PR solves it by simply correcting our counting of rate-limit-exceeded requests

**Comment by eseliger** [comment]
yep exactly, I should've noted that :)

File: enterprise/internal/completions/httpapi/limiter.go [hunk 1]
```diff
@@ -52,7 +54,7 @@ func (r *rateLimiter) TryAcquire(ctx context.Context) (err error) {
 		return errors.Wrap(err, "failed to read configured rate limit")
 	}
 
-	if limit == 0 {
+	if limit <= 0 {
```
Comment (unknwon) [inline enterprise/internal/completions/httpapi/limiter.go:57]: Love this

File: enterprise/internal/completions/httpapi/limiter.go [hunk 2]
```diff
@@ -84,13 +86,41 @@ func (r *rateLimiter) TryAcquire(ctx context.Context) (err error) {
...
+			// on every check, even if the limit was reached.
+			Used:       min(currentUsage, limit),
+			RetryAfter: time.Now().Add(time.Duration(ttl) * time.Second),
+		}
+	}
+
+	// Now that we know that we want to let the user pass, let's increment the rate
+	// limit counter for the user.
+	// Note that the rate limiter _may_ allow slightly more requests than the configured
+	// limit, incrementing the rate limit counter and reading the usage futher up are currently
```
Comment (unknwon) [inline enterprise/internal/completions/httpapi/limiter.go:119]:
```suggestion
	// limit, incrementing the rate limit counter and reading the usage further up are currently
```

Comment (unknwon) [reply enterprise/internal/completions/httpapi/limiter.go:119]: I think this type of compromise often has a simpler implementation than a super precise algorithm (which is often not required, practically speaking).

**Review by unknwon — APPROVED** [review]

**Pull request merged**